### ISSUE TITLE
`MultiLayerNetwork.output()` 20 times slower than `fit()` for LSTM

### ISSUE THREAD

**Issue by mzattera** [deeplearning4j/deeplearning4j#10168]
#### Issue Description

I have trained a LSTM network with 2 LSTM layers of 64 units and an output softmax layer (model attached: [LSTMModel.zip](https://github.com/user-attachments/files/18280588/LSTMModel.zip)).

During training, I use a data set with 5 million example sequences and a batch size of 128. An epoch lasts about 400 minutes on my machine, meaning an average of 5 millisecond to process a sequence, including *feed forward*, back propagation and evaluation vs. test set (of 100K samples).

I saved the model and I am now using it for predictions, but when I call `output()` it takes about 90 milliseconds on average for a single call; this is 15-20 times slower than the time it takes during training (and probably double that, considering during training also back propagation happens), so prediction (feed forward) is an order of magnitude slower than training.

I have put timing code wrapping the call to `output()`, so I am sure it is not due to ETL bottlenecks or other pieces of code (which, in any case, is the same used for training). Test code is attached: [Test.java.zip](https://github.com/user-attachments/files/18280578/Test.java.zip).

#### Version Information

I am using windows 10, with CUDA and cudNN, as confirmed by debug information:

```
10:07:17.345 [main] INFO  org.nd4j.linalg.factory.Nd4jBackend - Loaded [JCublasBackend] backend
10:07:19.400 [main] INFO  org.nd4j.nativeblas.NativeOpsHolder - Number of threads used for linear algebra: 32
10:07:19.441 [main] INFO  o.n.l.a.o.e.DefaultOpExecutioner - Backend used: [CUDA]; OS: [Windows 10]
10:07:19.441 [main] INFO  o.n.l.a.o.e.DefaultOpExecutioner - Cores: [12]; Memory: [4,0GB];
10:07:19.441 [main] INFO  o.n.l.a.o.e.DefaultOpExecutioner - Blas vendor: [CUBLAS]
10:07:19.466 [main] INFO  o.nd4j.linalg.jcublas.JCublasBackend - ND4J CUDA build version: 11.6.55
10:07:19.467 [main] INFO  o.nd4j.linalg.jcublas.JCublasBackend - CUDA device 0: [NVIDIA GeForce GTX 1070]; cc: [6.1]; Total memory: [8589803520]
10:07:19.468 [main] INFO  o.nd4j.linalg.jcublas.JCublasBackend - Backend build information:
 MSVC: 192930146
STD version: 201402L
DEFAULT_ENGINE: samediff::ENGINE_CUDA
HAVE_FLATBUFFERS
HAVE_CUDNN
```

Below extract of my pom.xml with configuration details:

```
	  <!-- Core -->
	  <dependency>
	      <groupId>org.deeplearning4j</groupId>
	      <artifactId>deeplearning4j-core</artifactId>
	      <version>1.0.0-M2.1</version>
	  </dependency>
       
		<!-- CUDA Engine (I have CUDA 11.8 and cudNN 9.6 installed on my machine-->
		<dependency>
		    <groupId>org.nd4j</groupId>
		    <artifactId>nd4j-cuda-11.6</artifactId>
		    <version>1.0.0-M2.1</version>
		</dependency>
		<dependency>
		    <groupId>org.nd4j</groupId>
		    <artifactId>nd4j-cuda-11.6</artifactId>
		    <version>1.0.0-M2.1</version>
		    <classifier>windows-x86_64-cudnn</classifier>
		</dependency>  
		<dependency>
		    <groupId>org.bytedeco</groupId>
		    <artifactId>cuda-platform-redist</artifactId>
		    <version>11.6-8.3-1.5.7</version>
		</dependency>
```

#### Additional Information

See attached model, test code, and dependencies.

**Comment by agibsonccc** [comment]
@mzattera how are you batching it? Training uses an AsyncDataSetIterator with precached batches. That means data is loaded in to the GPU in the background. I'd be more interested in a JMH benchmark that properly handles that case where you setup the problem first and load the GPU.

**Comment by mzattera** [comment]
Hi @agibsonccc, not sure if it helps, but for training I use a custom `DataSetIterator` that has `asyncSupported() = false`. Will it help understanding the problem if I do a test using a batch of size 128 (like I do in training) instead of a single sequence?

**Comment by mzattera** [comment]
In this test: [TestWBatches.java.zip](https://github.com/user-attachments/files/18281935/TestWBatches.java.zip) I use batches of size 128 (same as in training), on average it takes about 100-110 milliseconds a batch, meaning less than a millisecond per data point (faster than training, as expected). So, do I understand it correctly that slowness is due to the time it takes loading each single data point in the GPU memory, instead of loading an entire batch? ...or something like that?

**Comment by agibsonccc** [comment]
Loading data on the gpu is part of the time if it's not async. That's why I was asking you to run a JMH. If the difference is only during training then it's definitely because of the data loading because we do that transparently in training since we can make more assumptions.

**Comment by mzattera** [comment]
> That's why I was asking you to run a JMH.

Sorry @agibsonccc , but I have no idea what this is. Happy to do it if you clarify :)

**Comment by agibsonccc** [comment]
Just Google jmh harness. We have some samples in the
contriv/nd4j-benchmarks directory too.

**Comment by mzattera** [comment]
I repeated the test with the CPU backend and it takes, for single data points, 30ms instead of 90ms. This means using the CPU is three time faster than using the GPU when doing predictions on a single sample. 

Please notice that I use the LSTM to generate a sting, so when using it for predictions, I cannot use minibatches, as I have a single sequence that is built one char at a time.

I am NO EXPERT AT ALL with GPU and math libraries, but it seems strange a memory operation takes at least twice as long than a complex computation in the CPU...

**Comment by mzattera** [comment]
> Just Google jmh harness. We have some samples in the contriv/nd4j-benchmarks directory too.


OK, will give it a try when I have bit of time...thanks

**Comment by agibsonccc** [comment]
@mzattera that's fine. The main thing we need to do is confirm my theory about the data loading being the bottleneck and what happens if you let it warm up properly. If training works fine then the only difference would be how we handle the ETL since htat's what's built in.

**Comment by agibsonccc** [comment]
Closing — this has been addressed in the upcoming release (branch ag_new_release_updates_2). The relevant code has been rewritten/fixed.
